### ISSUE TITLE
Incorrect destination subnet for proxy request

### ISSUE THREAD

**Issue by JimG777** [davehorton/drachtio-server#25]
In a related question to issue #23, if the drachtio-server is listening on multiple network interfaces, say a host IP of 192.168.1.1 and a docker network IP of 172.22.0.1, the following happens in the proxy invite scenario using drachtio-srf:
```
A@172.22.0.2 -> 172.22.0.1 (client A sends INVITE to client B via the proxy)
172.22.0.1 -> 172.22.0.2 (100 Trying back to client A)
192.168.1.1 -> B@172.22.0.3 (proxy forwards INVITE to client B)
```
There is no routing from 192.168.1.1 to 172.22.0.3 so client B never sees the request.

Expectation: drachtio-server should be forwarding the INVITE using the 172.22.0.1 interface to a client on the 172.22.0.* subnet.

As a workaround I can limit the interfaces that drachtio-server listens on to just 172.22.0.1.

**Comment by davehorton** [comment]
I'm working on a fix for this, and would like to set up a scenario along the lines of what you have above.  However, I am not quite clear on how you configured things for your testing, could you help me understand?

It sounds like you created a docker network, and then have the drachtio-server running on the host machine but also somehow bound to the docker network.  I've not done that before, and when I create a docker network on my host (172.29.0.0/16) I don't see it in my `/sbin/ifconfig` output, thus I don't see how I can bind a drachtio-server instance on the host to that network.

Can you share the details of your setup so I can recreate?

**Comment by JimG777** [comment]
Hi Dave,

In my Docker environment, we use pre-defined Docker networks created with a command like:
docker network create test-network

After this, the network interface should show up in ifconfig. There should be a new bridge network in the list like "br-4984ea2882f".

If the drachtio-server is restarted, with the contact configuration set like "sip:*;transport=udp,tcp", the server will listen on the docker 'test-network' among others.

Finally, I would then run my SIP test clients inside Docker containers, associated with the created network.

If you aren't seeing the networks via ifconfig, it is strange. I am using Ubuntu 16.04 and Docker 17.12.0-ce.

Please let me know if I can help further. For now, this issue isn't high priority, since I have decided to run drachtio-server on a dedicated machine, so everything comes via the host address anyway.

**Comment by davehorton** [comment]
Hmm, I am running Docker 17.12.0-ce-mac45, so maybe that is my issue.  I've run into other anomolies running Docker on mac

I'll try to run the same test on a debian host.

For the record, here is what I tried on mac:
[dch-docker-network.txt](https://github.com/davehorton/drachtio-server/files/1636528/dch-docker-network.txt)

**Comment by JimG777** [comment]
It seems it is expected to be that way on the mac: https://docs.docker.com/engine/userguide/networking/#default-networks

There is a section
> Running on Docker for Mac or Docker for Windows?

I guess it won't show up as a network interface on mac or windows.

**Comment by davehorton** [comment]
Now I'm trying on Debian, and having a different issue.  I now see the docker network from the host and can bind drachtio to it, but when I send an INVITE from a container within that network the drachtio server on the host receives nothing.  Is there additional config you did?

Here my setup:

1. Host shows that it has an interface on the docker network
```
 ifconfig
docker0   Link encap:Ethernet  HWaddr 02:42:49:04:f9:2e  
          inet addr:172.17.0.1  Bcast:172.17.255.255  Mask:255.255.0.0
```
2. drachtio-server binds to it
```
./drachtio --contact "sip:*;transport=udp,tcp"
..
2018-01-16 15:30:04.857563 SipTransport::addTransports - creating transport: 0x1117d80: udp/172.17.0.1:5060
2018-01-16 15:30:04.857772 SipTransport::addTransports - creating transport: 0x1118230: tcp/172.17.0.1:5060
```

3. I start a sipp container on 172.17.0.3 in the docker network:
```
root@d23cef69a790:/# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
27: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.3/16 brd 172.17.255.255 scope global eth0
       valid_lft forever preferred_lft forever

sipp -m 1 -sn uac 172.17.0.1
```
but no sip packets reach my drachtio-server on the host?  Any thoughts?  Did you do something different?

**Comment by davehorton** [comment]
Belay that request....I figured out my issue: I had some iptables rules interfering with things.  Resolved that and I should be able to test your scenario now...

**Comment by davehorton** [comment]
I've checked in a fix for this on the cluster-experimental branch; I got my docker test environment working and it passed, but if you have a chance to test it in your environment as well to confirm at some point I'd be interested in your results.

**Comment by JimG777** [comment]
Thanks for the commit, I rebuilt to version g877ce69. Unfortunately I still see the issue.

Server is listening on 192.168.1.14 and 172.22.0.1.

```
Client A sends INVITE 172.22.0.5:5061 -> 172.22.0.1:5060 (UDP) - OK
Server sends 100 172.22.0.1:5060 ->172.22.0.5:5061 (UDP) - OK
Server sends INVITE to B 192.168.1.14:34608->172.22.0.2:5062 (TCP) - OK
Client B sends 180 172.22.0.2:5062 -> 192.168.1.14:34608 (TCP) - OK
Server sends 180 to A 192.168.1.14:5060 -> 172.22.0.5:5061 (UDP) - ICMP Failure
```
The ICMP failure is probably because UDP socket at client A is bound only to the 172.22.0.* interface. But still the Drachtio server socket is still choosing the Host interface for some reason.

**Comment by davehorton** [comment]
There are changes in the sofia-sip submodule for this, so after doing `git pull` did you also do `git submodule update --init --recursive` ?

**Comment by davehorton** [comment]
The sofia-sip submodule should be at `f2434608`

`f2434608 (HEAD -> master, origin/master, origin/HEAD) properly initialize array in prev checkin`

**Comment by JimG777** [comment]
You are correct, the dependency wasn't updated. I also had to remove the sofia library to make it build again (make clean didn't clear it). However there is still a problem. The interface Drachtio is choosing to send from is correct. But the destination IP is not correct. For example:
```
Client A sends INVITE 172.22.0.5:5061 -> 172.22.0.1:5060 (UDP) - OK
Server sends 100 172.22.0.1:5060 ->172.22.0.5:5061 (UDP) - OK
Server sends INVITE to B 172.22.0.1:56516->172.22.0.2:5062 (TCP) - Problem 1
Client B sends 180 172.22.0.2:5062 -> 172.20.0.1:56516 (TCP) - Problem 1
Server sends 180 to A 172.22.0.1:1024-> 172.22.0.5:5061 (UDP) - Problem 2
```
Problem 1: In the INVITE being forwarded to Client B, the server is adding a new via line like this:
`Via: SIP/2.0/TCP 172.20.0.1;rport;branch=z9hG4bK7c8ce60d-0df6-4672-8421-bb8605e3cfb6`

It is the wrong interface (one of the others that the server is listening on because it is started with the wildcard and there are a few other Docker networks on the host). Also there is no port. As a result, Client B sends the 180 to that interface (172.20.0.*) but at the same port I suppose via the existing TCP connection.

Problem 2: I am not sure if there is a client issue here, but for some reason I am getting ICMP error because the port is unreachable (172.22.0.5:5061), even though it was already handling UDP traffic. It is strange on the server side though, why is it using port 1024? We had successfully sent the 100 Trying from server 172.22.0.1:5060 to 172.22.0.5:5061 over UDP...

**Comment by davehorton** [comment]
Can you give me the steps to create the extra docker networks that you have, so I can recreate more closely your environment?  Till now I have just been starting docker and running my containers in the default 'bridge' network.

**Comment by JimG777** [comment]
You could make a few networks like this:
```
docker network create test1
docker network create test2
docker network create test3 --subnet 172.20.0.0/16 --gateway 172.20.0.1
docker network create drachtio-test --subnet 172.22.0.0/16 --gateway 172.22.0.1
```

The first two get default subnets, probably 172.18.0.0 and 172.19.0.0.

Then test3 is just a specific network and drachtio-test would be the one I use for clients.

I run my clients via docker-compose, with the following setting:
```
networks:
  default:
    external:
      name: drachtio-test
```

Then, any services I create will use this network by default. For example:
```
services:
  some_client_1
    image: "some_image"
```

When some_client_1 runs, it will be assigned an IP from the 172.22.0.* subnet.

In my case, I run a couple of the clients like this and drachtio is listening on all interfaces, including the host. One client will send an INVITE to the other via drachtio-server, which will attempt to proxy to the other client.

To clean up:
`docker network rm test1 test2 test3 drachtio-test`

**Comment by davehorton** [comment]
OK, take 3!  can you try again with the latest set of checkins?  Again, you will need to update the submodules. (I made `make clean` clean all dependencies).

I think I have the same set up now as you, and it was working for me.  Let me know how you get on.

**Comment by JimG777** [comment]
Hi Dave, it is working smoothly now! I tested back and forth and with an external client and the interfaces are all exactly correct. Thanks also for the `clean` adjustment.

You can probably close this now.

_BTW, I re-submitted pull requests for drachtio & drachtio-srf. It seems npm is pulling from the develop branches._

Thanks for all the work on this one!
